### PR TITLE
feat: pushdown get_format into TiFlash

### LINKED ISSUE #5115: Implement `GetFormat` function push down
## Feature Request

**Describe the feature you'd like:**

We'd like to support `GetFormat` function push down.

**Teachability, Documentation, Adoption, Migration Strategy:**

At least one integration test and several unit tests.

### PR DESCRIPTION
**Pull request by wirybeaver**
### What problem does this PR solve?

Issue Number: close #5115 
Related TiDB Issue: pingcap/tidb#36035

Problem Summary:
Implement GetFormat push down. Refer to the rule in the [MySQL doc](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_str-to-date)

Trigger manual integration test with the following command
```
➜  ~tiup playground --tiflash.binpath /Users/xuanyili/my-code/tiflash/build/dbms/src/Server/tiflash --db.binpath /Users/xuanyili/my-code/tidb/bin/tidb-server
```

The explain result shows that the expression `get_format` will be pushdown to TiFlash.
```
+---------------------------+----------+--------------+---------------+---------------------------------------------+
| id                        | estRows  | task         | access object | operator info                               |
+---------------------------+----------+--------------+---------------+---------------------------------------------+
| TableReader_9             | 10000.00 | root         |               | data:ExchangeSender_8                       |
| └─ExchangeSender_8        | 10000.00 | mpp[tiflash] |               | ExchangeType: PassThrough                   |
|   └─Projection_4          | 10000.00 | mpp[tiflash] |               | get_format(DATE, test.t.location)->Column#3 |
|     └─TableFullScan_7     | 10000.00 | mpp[tiflash] | table:t       | keep order:false, stats:pseudo              |
+---------------------------+----------+--------------+---------------+---------------------------------------------+
```
Unit Test
```
ninja gtests_dbms
./dbms/gtests_dbms --gtest_filter='GetFormatTest*'"
```
Integration Test
```
./run-test.sh fullstack-test/expr/get_format.test
```

### What is changed and how it works?
- The first parameter is a MySQL reserved word without quote in the set of (`DATETIME`, `DATE`, `TIME`, `TIMESTAMP`). When the argument is push down to TiFlash, it should always be a ColumnConst. Thus, overwrite `getArgumentsThatAreAlwaysConstant()` by returning `{0}`.
- returns empty string in case of no rule matching according to the [TiDB implementation]((https://sourcegraph.com/github.com/pingcap/tidb@master/-/blob/expression/builtin_time_vec.go?L621)).

### Check List

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility
### Release note

```release-note
Support get_format pushdown to TiFlash.
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1811]: This is the default implementation and does not need to be overridden.

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1816]:
```suggestion
            throw Exception("First argument for function " + getName() + " must be String", ErrorCodes::ILLEGAL_COLUMN);
```
Here only checks whether it is string type

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1818]: ditto

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1824]: add comment that todo support non-constant time type?

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1829]: add `assert(location_col)`

File: dbms/src/Functions/FunctionsConversion.h
Comment (SeaRise) [inline dbms/src/Functions/FunctionsConversion.h:1757]:
```suggestion
    static String get_format(const StringRef & time_type, const StringRef & location)
```
And then we don't need to call `StringRef.toString()`

**Comment by SeaRise** [comment]
And add UT under [dbms/src/Functions/tests](https://github.com/pingcap/tiflash/tree/a89222abd6ad1934c54838531532a1253bf4f66c/dbms/src/Functions/tests)

**Comment by SeaRise** [comment]
Add release note like 'Support to pushdown get_format to TiFlash'.
And add the pr link of tidb?

**Comment by wirybeaver** [comment]
> And add UT under [dbms/src/Functions/tests](https://github.com/pingcap/tiflash/tree/a89222abd6ad1934c54838531532a1253bf4f66c/dbms/src/Functions/tests)


> And add UT under [dbms/src/Functions/tests](https://github.com/pingcap/tiflash/tree/a89222abd6ad1934c54838531532a1253bf4f66c/dbms/src/Functions/tests)

Done

**Comment by wirybeaver** [comment]
> Add release note like 'Support to pushdown get_format to TiFlash'. And add the pr link of tidb?

Done

**Comment by wirybeaver** [comment]
Rebase master instead of Merge. However, encounter a build error. Seems to a lack of contrib library.
```
CMake Error at contrib/prometheus-cpp-cmake/pull/CMakeLists.txt:14 (add_library):
  Cannot find source file:
    /Users/xuanyili/my-code/tiflash/contrib/prometheus-cpp/pull/src/basic_auth.cc
  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
CMake Error at contrib/prometheus-cpp-cmake/push/CMakeLists.txt:5 (add_library):
  Cannot find source file:
    /Users/xuanyili/my-code/tiflash/contrib/prometheus-cpp/push/src/curl_wrapper.cc
  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
CMake Error at contrib/prometheus-cpp-cmake/pull/CMakeLists.txt:14 (add_library):
  No SOURCES given to target: pull
CMake Error at contrib/prometheus-cpp-cmake/push/CMakeLists.txt:5 (add_library):
  No SOURCES given to target: push
```

**Comment by SeaRise** [comment]
> Add release note like 'Support to pushdown get_format to TiFlash'. And add the pr link of tidb?

sorry, i meant to post the tidb pr in the pr description.:)

**Comment by wirybeaver** [comment]
> > Add release note like 'Support to pushdown get_format to TiFlash'. And add the pr link of tidb?
> 
> sorry, I meant to post the tidb pr in the pr description.:)

No worries!

**Comment by wirybeaver** [comment]
rebase

**Comment by wirybeaver** [comment]
@SeaRise `1 workflow awaiting approval` Could you approve the workflow? The `/run-all-tests` doesn't trigger the test. I have run unit test and integration test locally and everything looks good.

**Comment by wirybeaver** [comment]
> Rebase master instead of Merge. However, encounter a build error. Seems to a lack of contrib library.
> 
> ```
> CMake Error at contrib/prometheus-cpp-cmake/pull/CMakeLists.txt:14 (add_library):
>   Cannot find source file:
>     /Users/xuanyili/my-code/tiflash/contrib/prometheus-cpp/pull/src/basic_auth.cc
>   Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
>   .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
> CMake Error at contrib/prometheus-cpp-cmake/push/CMakeLists.txt:5 (add_library):
>   Cannot find source file:
>     /Users/xuanyili/my-code/tiflash/contrib/prometheus-cpp/push/src/curl_wrapper.cc
>   Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
>   .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
> CMake Error at contrib/prometheus-cpp-cmake/pull/CMakeLists.txt:14 (add_library):
>   No SOURCES given to target: pull
> CMake Error at contrib/prometheus-cpp-cmake/push/CMakeLists.txt:5 (add_library):
>   No SOURCES given to target: push
> ```

It's my bad. I should have run `git submodule update --init --recursive` after rebasing.

**Comment by wirybeaver** [comment]
the integration test sample.test always have a timeout error:
```
[2022-07-05T22:02:49.327Z]     => waiting for test.t available
[2022-07-05T22:02:49.327Z]     => waiting for test.t available
[2022-07-05T22:02:49.327Z]     => cannot sync tables in 600.023127794s
[2022-07-05T22:02:49.327Z]   Expected:
[2022-07-05T22:02:49.327Z]     <nothing> 
```

**Comment by wirybeaver** [comment]
@SeaRise Assume the sample.test timeout issue would be solved, shall I merge the related TiDB PR [35950](https://github.com/pingcap/tidb/pull/35950) before triggering the integration test?

**Comment by wirybeaver** [comment]
Get the same time out exception when verify sample.test

**Comment by wirybeaver** [comment]
> /run-integration-test

the integration test sample.test always has a timeout error:
```
[2022-07-06T00:33:33.530Z] + docker-compose -f cluster.yaml -f tiflash-dt.yaml exec -T tiflash0 bash -c 'cd /tests ; ./run-test.sh fullstack-test'
[2022-07-06T00:43:40.288Z]   File: fullstack-test/sample.test
[2022-07-06T00:43:40.288Z]   Error line: 22
[2022-07-06T00:43:40.288Z]   Error: wait_table test t
[2022-07-06T00:43:40.288Z]   Result:
[2022-07-06T00:43:40.288Z]     => wait for test.t available in TiFlash
[2022-07-06T00:43:40.288Z]     => waiting for test.t available
[2022-07-05T22:02:49.327Z]     => waiting for test.t available
[2022-07-05T22:02:49.327Z]     => waiting for test.t available
[2022-07-05T22:02:49.327Z]     => cannot sync tables in 600.023127794s
[2022-07-05T22:02:49.327Z]   Expected:
[2022-07-05T22:02:49.327Z]     <nothing> 
```

**Comment by SeaRise** [comment]
> @SeaRise Assume the sample.test timeout issue would be solved, shall I merge the related TiDB PR [35950](https://github.com/pingcap/tidb/pull/35950) before triggering the integration test? I am concerned the get_format integration test will fail if the TiDB doesn't support pushdown to TiFlash

Don't worry, integration tests will run even if TiDB PR hasn't been merged.
Because the execution result of `get_format ` in `TiDB` and `TiFlash` is the same

**Comment by wirybeaver** [comment]
@SeaRise 
I tried to reproduce the timeout issue of sample.test at my Mac laptop and got a different error:
```
➜  tests git:(xuanyili/getFormat) ./run-test.sh fullstack-test/sample.test         
  File: fullstack-test/sample.test
  Error line: 24
  Error: set session tidb_isolation_read_engines='tiflash'; select s, i from test.t
  Result:
    ERROR 1105 (HY000) at line 1: lastRange in paging should not be nil
```
I randomly picked up some other expression integration tests and got passed. To me, this `lastRange` error is a distracting problem, which might be arise from by MacOS and Linux difference. 

Anyway, do you have any insights about the [timeout issue](https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-integration-tests/detail/tiflash-ghpr-integration-tests/7336/pipeline) of table syncing up? I suspected it's caused by the network broken among docker containers.

**Comment by SeaRise** [comment]
> I randomly picked up some other expression integration tests and got passed. To me, this `lastRange` error is a distracting problem, which might be arise from by MacOS and Linux difference.
> 
> Anyway, do you have any insights about the [timeout issue](https://ci.pingcap.net/blue/organizations/jenkins/tiflash-ghpr-integration-tests/detail/tiflash-ghpr-integration-tests/7336/pipeline) of table syncing up? I suspected it's caused by the network broken among docker containers.

Don't worry, this is a problem on the pd side, someone is handling it:)

File: dbms/src/Functions/FunctionsConversion.h
Comment (zanmato1984) [inline dbms/src/Functions/FunctionsConversion.h:1820]: I don't think we should consider a non-const first column for this case.

`get_format` is designed to have the first argument as a literal - note it's even not only a literal, but also a keyword (no quote):

```
mysql> select get_format("date", "USA");
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '"date", "USA")' at line 1
```

Comment (wirybeaver) [reply dbms/src/Functions/FunctionsConversion.h:1820]: Thanks for the clarification. I would change the comment saying: 
`The first argument is designed as a MySQL reserved word. You would even encounter a syntax error when wrap it around with quote in SQL. For example, select GET_FORMAT("DATE", "USA") will fail. Removing the quote can solve the problem. Thus the first argument should always be a ColumnConst.`

**Comment by wirybeaver** [comment]
@zanmato1984 Change the comment about the reason why first argument MUST be ColumnConst. Also rebase master. Please check it.

**Comment by zanmato1984** [comment]
> `git rebase` OR `git merge`? Right now I use `git commit --amend` to keep my changes as a single commit and use `git pull --rebase` to fetch remote master change.

You are right. We don't have particular workflow for private branches. Squash merging into master is all.

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Comment by zanmato1984** [comment]
Nice work and thanks for you contribution!

**Comment by wirybeaver** [comment]
> @wirybeaver Nice work and thanks for your contribution!

Thanks for your guys comments and suggestions! Also a shout to @SeaRise

**Comment by wirybeaver** [comment]
@SeaRise Could you /merge this branch when you are available?

**Pull request merged**
